### PR TITLE
cluster_staypoints: add support for big data

### PR DESCRIPTION
**Pull request by hong2223**
- change the `cluster_staypoints `function: the DBSCAN now directly accepts the distance metric and does not use a pre-calculated pairwise distance matrix.
- add test for haversine distance DBSCAN
- update author information

### CODE REVIEW

**Comment by hong2223** [comment]
I am not sure if I should add my name in the **copyright**, please check

**Review by henrymartin1 — CHANGES_REQUESTED** [review]
address comments before merging

**Comment by henrymartin1** [comment]
> 
> 
> address comments before merging

(This concerns only the comment about the epsilon, the others are rather general...)

File: trackintel/preprocessing/staypoints.py [hunk 1]
```diff
@@ -59,24 +58,21 @@ def cluster_staypoints(staypoints,
     ret_sp = staypoints.copy()
     if method=='dbscan':
 
-        if distance_matrix_metric is not None:
-            db = DBSCAN(eps=epsilon, min_samples=num_samples, metric='precomputed')
-        else:    
-            db = DBSCAN(eps=epsilon, min_samples=num_samples)
+        if distance_matrix_metric == 'haversine':
+            epsilon = epsilon / 6371000 # convert to radius
+        db = DBSCAN(eps=epsilon, min_samples=num_samples, algorithm='ball_tree', metric=distance_matrix_metric)
```
[the remark this answers is not in the document]
Comment (hong2223) [reply trackintel/preprocessing/staypoints.py:63]: the R-tree method does not support 'haversine' distance calculation

File: trackintel/preprocessing/staypoints.py [hunk 2]
```diff
@@ -59,24 +58,21 @@ def cluster_staypoints(staypoints,
     ret_sp = staypoints.copy()
     if method=='dbscan':
 
-        if distance_matrix_metric is not None:
-            db = DBSCAN(eps=epsilon, min_samples=num_samples, metric='precomputed')
-        else:    
-            db = DBSCAN(eps=epsilon, min_samples=num_samples)
+        if distance_matrix_metric == 'haversine':
```
[the remark this answers is not in the document]
Comment (hong2223) [reply trackintel/preprocessing/staypoints.py:61]: the input and output of scikit-learn’s haversine metric are both in radian units, see [clustering example](https://geoffboeing.com/2014/08/clustering-to-reduce-spatial-data-set-size/) and our [own code example ](https://github.com/mie-lab/trackintel/blob/master/tests/geogr/test_distances.py). The `epsilon` here is directly applied to the metric output, thus needs to be in radian.

File: trackintel/preprocessing/staypoints.py [hunk 3]
```diff
@@ -59,24 +58,21 @@ def cluster_staypoints(staypoints,
...
             for user_id_this in ret_sp["user_id"].unique():
                 # Slice staypoints array by user. This is not a copy!
                 user_staypoints = ret_sp[ret_sp["user_id"] == user_id_this]  
                 
-                if distance_matrix_metric is not None:
-                    sp_distance_matrix = calculate_distance_matrix(user_staypoints, 
-                                                                   dist_metric=distance_matrix_metric)
-                    labels = db.fit_predict(sp_distance_matrix)
+                if distance_matrix_metric == 'haversine':
+                    p = np.array([[radians(g.y), radians(g.x)] for g in user_staypoints.geometry])
```
[the remark this answers is not in the document]
Comment (hong2223) [reply trackintel/preprocessing/staypoints.py:75]: comments added

**Review by henrymartin1 — APPROVED** [review]

**Comment by henrymartin1** [comment]
Nice! Thanks for the clarification.

**Pull request merged**